File: src/main/java/com/sgx/icms/service/AuthService.java
package com.sgx.icms.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.sgx.icms.dao.JdbcUserDao;
import com.sgx.icms.dao.UserDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.domain.AuditLog;
import com.sgx.icms.domain.User;

/**
 * Authentication use-case: verify credentials, enforce account status, record the
 * login attempt, and stamp {@code last_login}. Returns the authenticated {@link User}
 * or {@code null} — callers must not be able to distinguish "no such user" from
 * "wrong password" (no user enumeration).
 */
public class AuthService {

    private static final Logger LOG = LoggerFactory.getLogger(AuthService.class);

    private final UserDao userDao = new JdbcUserDao();
    private final PasswordService passwords = new PasswordService();
    private final AuditService audit = new AuditService();

    /**
     * @return the authenticated user, or {@code null} if credentials are invalid or
     *         the account is not ACTIVE.
     */
    public User authenticate(String username, String rawPassword, String ip) {
        if (username == null || username.isEmpty() || rawPassword == null || rawPassword.isEmpty()) {
            return null;
        }
        final String uname = username.trim();

        User user = Db.withConnection(conn -> userDao.findByUsername(conn, uname));

        boolean ok = user != null
                && user.isActive()
                && passwords.matches(rawPassword, user.getPasswordHash());

        if (!ok) {
            // Don't reveal which check failed; log internally for ops.
            LOG.info("Failed login for username='{}' from {}", uname, ip);
            audit.record(user == null ? null : user.getId(), uname,
                    user == null ? null : user.getRoleName(),
                    "LOGIN_FAIL", "username:" + uname, AuditLog.RESULT_FAILED, ip);
            return null;
        }

        final long uid = user.getId();
        Db.inTransaction(conn -> {
            userDao.updateLastLogin(conn, uid);
            return null;
        });
        audit.record(user.getId(), user.getUsername(), user.getRoleName(),
                "LOGIN", "user:" + user.getUsername(), AuditLog.RESULT_SUCCESS, ip);
        LOG.info("User '{}' ({}) logged in from {}", user.getUsername(), user.getRoleName(), ip);
        return user;
    }
}

File: backend/src/services/authService.js
'use strict';

const { knex } = require('../db/knex');
const { withTransaction } = require('../db/tx');
const userRepo = require('../repositories/userRepo');
const passwords = require('./passwordService');
const audit = require('./auditService');
const jwtUtil = require('../utils/jwt');
const logger = require('../config/logger');

/**
 * Authentication use-case — Node port of AuthService. Verifies credentials,
 * enforces ACTIVE status, records the attempt, stamps last_login, and (on success)
 * issues a JWT. Callers cannot distinguish "no such user" from "wrong password".
 */

function toPublicUser(u) {
  return {
    id: Number(u.id),
    username: u.username,
    fullName: u.fullName,
    email: u.email,
    role: u.roleName,
    branch: u.branch,
  };
}

/**
 * @returns {Promise<{token, user}|null>} null when invalid or not ACTIVE.
 */
async function authenticate(username, rawPassword, ip) {
  if (!username || !rawPassword) return null;
  const uname = String(username).trim();

  const user = await userRepo.findByUsername(knex, uname);

  const ok =
    user &&
    user.status === 'ACTIVE' &&
    (await passwords.matches(rawPassword, user.passwordHash));

  if (!ok) {
    logger.info({ username: uname, ip }, 'Failed login');
    await audit.recordSafe({
      userId: user ? Number(user.id) : null,
      username: uname,
      role: user ? user.roleName : null,
      action: 'LOGIN_FAIL',
      entity: `username:${uname}`,
      result: 'FAILED',
      ip,
    });
    return null;
  }

  await withTransaction((trx) => userRepo.updateLastLogin(trx, user.id));
  await audit.recordSafe({
    userId: Number(user.id),
    username: user.username,
    role: user.roleName,
    action: 'LOGIN',
    entity: `user:${user.username}`,
    result: 'SUCCESS',
    ip,
  });
  logger.info({ username: user.username, role: user.roleName, ip }, 'User logged in');

  const publicUser = toPublicUser(user);
  return { token: jwtUtil.sign({ ...publicUser, roleName: publicUser.role }), user: publicUser };
}

/** Logout is stateless (client drops the token); we only record the audit row. */
async function logout(actor, ip) {
  if (actor) {
    await audit.success(actor, 'LOGOUT', `user:${actor.username}`, ip);
  }
}

module.exports = { authenticate, logout, toPublicUser };
